### ISSUE TITLE
ArrayIndexOutOfBoundsException HtmlResponseWriter.writeUnescapedCData

### ISSUE THREAD

**Issue by tkernstock** [payara/Payara#4588]
# Description #
I would like to change from Payara 5.184 to 5.201. All 3 deployed Webapplications are working fine on 5.184 and also on 5.201 with one exception.
 
I have a searchpage for jobs that works with partial rendering, updating only the job section on new searches. This can be found under https://www.isg.com/jobs/search/index.jsf. and works fine on 5.184. 

In 5.201 the page isn't refreshed and I find an error in the logfile:

`java.lang.ArrayIndexOutOfBoundsException: -2
	at com.sun.faces.renderkit.html_basic.HtmlResponseWriter.writeUnescapedCData(HtmlResponseWriter.java:1474)
	at com.sun.faces.renderkit.html_basic.HtmlResponseWriter.writeText(HtmlResponseWriter.java:961)
	at javax.faces.context.ResponseWriter.writeText(ResponseWriter.java:340)
	at javax.faces.context.ResponseWriterWrapper.writeText(ResponseWriterWrapper.java:312)
	at javax.faces.context.ResponseWriterWrapper.writeText(ResponseWriterWrapper.java:312)
	at javax.faces.context.ResponseWriterWrapper.writeText(ResponseWriterWrapper.java:312)
	at javax.faces.context.ResponseWriterWrapper.writeText(ResponseWriterWrapper.java:312)
	at com.sun.faces.renderkit.html_basic.OutcomeTargetLinkRenderer.writeValue(OutcomeTargetLinkRenderer.java:159)
	at com.sun.faces.renderkit.html_basic.OutcomeTargetLinkRenderer.renderAsActive(OutcomeTargetLinkRenderer.java:140)
	at com.sun.faces.renderkit.html_basic.OutcomeTargetLinkRenderer.encodeBegin(OutcomeTargetLinkRenderer.java:72)
	at javax.faces.component.UIComponentBase.encodeBegin(UIComponentBase.java:540)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1644)
	at javax.faces.render.Renderer.encodeChildren(Renderer.java:152)
	at javax.faces.component.UIComponentBase.encodeChildren(UIComponentBase.java:566)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1647)
	at com.sun.faces.context.PartialViewContextImpl$PhaseAwareVisitCallback.visit(PartialViewContextImpl.java:615)
	at com.sun.faces.component.visit.PartialVisitContext.invokeVisitCallback(PartialVisitContext.java:159)
	at javax.faces.component.UIForm.visitTree(UIForm.java:365)
	at javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at javax.faces.component.UIComponent.visitTree(UIComponent.java:1468)
	at com.sun.faces.context.PartialViewContextImpl.processComponents(PartialViewContextImpl.java:400)
	at com.sun.faces.context.PartialViewContextImpl.processPartial(PartialViewContextImpl.java:319)
	at org.primefaces.context.PrimePartialViewContext.processPartial(PrimePartialViewContext.java:63)
	at javax.faces.context.PartialViewContextWrapper.processPartial(PartialViewContextWrapper.java:228)
	at org.omnifaces.context.OmniPartialViewContext.processPartial(OmniPartialViewContext.java:122)
	at javax.faces.component.UIViewRoot.encodeChildren(UIViewRoot.java:1102)
	at javax.faces.component.UIComponent.encodeAll(UIComponent.java:1647)
	at com.sun.faces.application.view.FaceletViewHandlingStrategy.renderView(FaceletViewHandlingStrategy.java:442)
	at com.sun.faces.application.view.MultiViewHandler.renderView(MultiViewHandler.java:170)
	at javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:132)
	at javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:132)
	at javax.faces.application.ViewHandlerWrapper.renderView(ViewHandlerWrapper.java:132)
	at org.omnifaces.viewhandler.OmniViewHandler.renderView(OmniViewHandler.java:121)
	at com.sun.faces.lifecycle.RenderResponsePhase.execute(RenderResponsePhase.java:102)
	at com.sun.faces.lifecycle.Phase.doPhase(Phase.java:76)
	at com.sun.faces.lifecycle.LifecycleImpl.render(LifecycleImpl.java:199)
	at org.apache.deltaspike.jsf.impl.listener.request.DeltaSpikeLifecycleWrapper.render(DeltaSpikeLifecycleWrapper.java:111)
	at javax.faces.lifecycle.LifecycleWrapper.render(LifecycleWrapper.java:88)
	at org.apache.deltaspike.jsf.impl.listener.request.JsfClientWindowAwareLifecycleWrapper.render(JsfClientWindowAwareLifecycleWrapper.java:160)
	at javax.faces.webapp.FacesServlet.executeLifecyle(FacesServlet.java:708)
	at javax.faces.webapp.FacesServlet.service(FacesServlet.java:451)
	at org.apache.catalina.core.StandardWrapper.service(StandardWrapper.java:1636)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:331)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:211)
	at org.omnifaces.filter.FacesExceptionFilter.doFilter(FacesExceptionFilter.java:133)
	at org.omnifaces.filter.FacesExceptionFilter.doFilter(FacesExceptionFilter.java:118)
	at org.omnifaces.filter.HttpFilter.doFilter(HttpFilter.java:108)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:253)
	at .......

## Context (Optional)
Since the search parameters on the page are saved in the session, klicking another link and then returning to the searchpage, which is then rendered completely, shows the correct results.

What I found searching for a solution was, that this problem seems to be Mojarra based. Like:
https://github.com/eclipse-ee4j/mojarra/issues/4650
https://github.com/jboss/mojarra/pull/56
https://developer.jboss.org/thread/281044

## Environment ##

- **Payara Version**: 5.201
- **Edition**: Full
- **JDK Version**: 1.8.0_241 Oracle
- **Operating System**: Windows10
- **Database**: MariaDB

**Comment by fturizo** [comment]
@tkernstock, can you please provide a self-contained reproducer that allow us to verify the issue you are experiencing at the moment? If you can upload the reproducer or link to a public online repository, we can provide further assistance.

**Comment by tkernstock** [comment]
I will try. But the project with this problem is rather complex and needs a lot of data from a database, so I not sure if I can do that in a justifiable timeframe.

**Comment by fturizo** [comment]
@tkernstock, you do not need to prepare the entire application to be shared, we only need a **Short Self Contained Example** (more info here: http://sscce.org/).  You only need to provide a short example where the JSF error product of the partial rendering is shown. 

If the search function is needed to trigger the error, just ship the application with dummy data instead.

**Comment by tkernstock** [comment]
the page is composed with templates and 8 includes all with different backing beans + database access. so the "short" won't be short. I need to talk to my customer if he is willing to pay my time creating this testcase.

**Comment by khasunuma** [comment]
I received an e-mail from a Japanese community person who cannot read/write English. He reported almost same problem in the mail. He cannot provide a reproducer sample as he is too busy now, but I believe this issue would be occurred generally if an application use with PrimeFaces on Payara Server 5.201.

As the related issues provided by @tkernstock, it's seemed a bug of Mojarra 2.3.14, however the Mojarra project probably focuses at developing Mojarra 3.0.0 for Jakarta EE 9. Therefore, Payara needs to create a patch build for Mojarra 2.3.14 to resolve the issue.

**Comment by khasunuma** [comment]
@tkernstock as I look this and relevant issues, this problem might be occurred by smaller examples without database access. Would you please try to create a simple reproducer to resolve the issue faster if you have a time? I apologised that I currently cannot create a reproducer because I have no information about your application(s) without you use JSF with PrimeFaces.

**Comment by tkernstock** [comment]
@khasunuma: I will try creating a reproducer, but it may take a while.

**Comment by tkernstock** [comment]
@khasunuma: I tried to create a short reproducer for a couple of hours now, removing all DB access and templates etc., but it seems I removed too much. Since I don't know what's the cause of the error I don't know what is essential to leave in the testcode. Sorry, but there is no way to do that in an exeptable  timeframe. 

What I can say though, that in the method of Mojarra that throws the error, the boundaries of the arrays that are worked with are not checked. So it's no wonder that a ArrayOutOfBounds is thrown. 
Another possibility would be to check what the jboss solution was to the problem. (see my former post)

**Comment by tkernstock** [comment]
@khasunuma: I finnaly found the problem, by stripping code from my page. The error in Myfaces was caused by the existance of a Javascriptcode. Allthough the code produced correctr XHTML when called, the simple existance caused the problem. I simply moved the code to a .js file and now everthing works fine.

**Comment by Jarannn** [comment]
Why is this closed when it was solved by a workaround that works only for a few? The issue has however been resolved for me by using the HtmlResponseWriter from the release originally posted by @mattwltr in https://github.com/eclipse-ee4j/mojarra/issues/4647#issuecomment-591341704_
I think the following pull request solved the issue: https://github.com/jboss/mojarra/pull/57

**Comment by tkernstock** [comment]
@Jarannn:  Well the answer is easy, I spent hours trying to create a reproducer which I couldn't manage. Although I mentioned the problem in Mojarra, it seemed that nowbody cares about this bug without a reproducer, so I closed the case. Feel free to reopen it.

**Comment by Jarannn** [comment]
> 
> 
> @Jarannn: Well the answer is easy, I spent hours trying to create a reproducer which I couldn't manage.

There's a reproducable example in https://github.com/eclipse-ee4j/mojarra/issues/4647 (haven't tried it myself but it's fundamentally the same issue as in my code).
@fturizo has just removed the reproducer-needed label, maybe because of the example above? How can one re-open this issue? Sorry I'm new to GitHub. Would this be fixed within the Payara project itself anyway or only after it has been fixed upstream?

**Comment by fturizo** [comment]
@Jarannn, as the original requester of this issue explained, we require a test reproducer to verify the bug report you are filing. Without one, I'm afraid that we can't offer much assistance. Since the issue was closed by @tkernstock, we removed the label since the issue no longer requires any resolution from us. If you are still interested in implementing a fix, you can raise a new issue and provide a clear reproducer that evidences the error in release *5.201*.
